// @tamagui/themes/dist/cjs/v2.native.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
}, __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var v2_exports = {};
__export(v2_exports, {
  blue: () => import_colors.blue,
  blueDark: () => import_colors.blueDark,
  gray: () => import_colors8.gray,
  grayDark: () => import_colors8.grayDark,
  green: () => import_colors2.green,
  greenDark: () => import_colors2.greenDark,
  orange: () => import_colors4.orange,
  orangeDark: () => import_colors4.orangeDark,
  pink: () => import_colors6.pink,
  pinkDark: () => import_colors6.pinkDark,
  purple: () => import_colors7.purple,
  purpleDark: () => import_colors7.purpleDark,
  red: () => import_colors3.red,
  redDark: () => import_colors3.redDark,
  yellow: () => import_colors5.yellow,
  yellowDark: () => import_colors5.yellowDark
});
module.exports = __toCommonJS(v2_exports);
__reExport(v2_exports, require("./tokens"), module.exports);
__reExport(v2_exports, require("./masks"), module.exports);
__reExport(v2_exports, require("./componentThemeDefinitions"), module.exports);
__reExport(v2_exports, require("./palettes"), module.exports);
__reExport(v2_exports, require("./templates"), module.exports);
__reExport(v2_exports, require("./shadows"), module.exports);
var import_colors = require("@tamagui/colors"), import_colors2 = require("@tamagui/colors"), import_colors3 = require("@tamagui/colors"), import_colors4 = require("@tamagui/colors"), import_colors5 = require("@tamagui/colors"), import_colors6 = require("@tamagui/colors"), import_colors7 = require("@tamagui/colors"), import_colors8 = require("@tamagui/colors");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  blue,
  blueDark,
  gray,
  grayDark,
  green,
  greenDark,
  orange,
  orangeDark,
  pink,
  pinkDark,
  purple,
  purpleDark,
  red,
  redDark,
  yellow,
  yellowDark,
  ...require("./tokens"),
  ...require("./masks"),
  ...require("./componentThemeDefinitions"),
  ...require("./palettes"),
  ...require("./templates"),
  ...require("./shadows")
});
