// @tamagui/static/dist/cjs/registerRequire.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var registerRequire_exports = {};
__export(registerRequire_exports, {
  getNameToPaths: () => getNameToPaths,
  registerRequire: () => registerRequire,
  setRequireResult: () => setRequireResult
});
module.exports = __toCommonJS(registerRequire_exports);
var import_node = require("esbuild-register/dist/node"), import_requireTamaguiCore = require("./helpers/requireTamaguiCore");
const nameToPaths = {}, Module = require("module"), getNameToPaths = () => nameToPaths, proxyWorm = require("@tamagui/proxy-worm"), rnw = require("react-native-web");
let isRegistered = !1, og;
const whitelisted = {
  react: !0
}, compiled = {};
function setRequireResult(name, result) {
  compiled[name] = result;
}
function registerRequire(platform, { proxyWormImports } = {
  proxyWormImports: !1
}) {
  if (isRegistered)
    return {
      tamaguiRequire: require,
      unregister: () => {
      }
    };
  const { unregister } = (0, import_node.register)({
    hookIgnoreNodeModules: !1
  });
  og || (og = Module.prototype.require), isRegistered = !0, Module.prototype.require = tamaguiRequire;
  function tamaguiRequire(path) {
    if (path === "tamagui")
      return og.apply(this, ["tamagui/native"]);
    if (path === "@tamagui/core" || path === "@tamagui/web")
      return (0, import_requireTamaguiCore.requireTamaguiCore)(platform, (path2) => og.apply(this, [path2]));
    if (path in compiled)
      return compiled[path];
    if (/\.(gif|jpe?g|png|svg|ttf|otf|woff2?|bmp|webp)$/i.test(path))
      return {};
    if (path === "@gorhom/bottom-sheet" || path.startsWith("react-native-reanimated") || path === "expo-linear-gradient" || path === "@expo/vector-icons" || path === "tamagui/linear-gradient" || path === "react-native-svg")
      return proxyWorm;
    if (path === "react-native/package.json")
      return packageJson;
    if (path === "react-native-web-lite" || path.startsWith("react-native"))
      return rnw;
    if (path in knownIgnorableModules)
      return proxyWorm;
    if (!whitelisted[path] && proxyWormImports && !path.includes(".tamagui-dynamic-eval"))
      return path === "tamagui" ? og.apply(this, [path]) : proxyWorm;
    try {
      return og.apply(this, arguments);
    } catch (err) {
      return !process.env.TAMAGUI_ENABLE_WARN_DYNAMIC_LOAD && path.includes("tamagui-dynamic-eval") ? void 0 : (allowedIgnores[path] || IGNORES === "true" || (process.env.TAMAGUI_SHOW_FULL_BUNDLE_ERRORS ? console.error(
        `Tamagui failed loading "${path}"
  
  ${err.message}
  ${err.stack}

  `
      ) : hasWarnedForModules.has(path) || (hasWarnedForModules.add(path), console.log(
        `
\u26A0\uFE0F Tamagui Warning [001]: Skipping loading ${path} due to error bundling.
   - message: ${err.message}
   - for more info see: https://tamagui.dev/docs/intro/errors#warning-001
   - set TAMAGUI_SHOW_FULL_BUNDLE_ERRORS=1 to see stack trace

`
      ))), proxyWorm);
    }
  }
  return {
    tamaguiRequire,
    unregister: () => {
      unregister(), isRegistered = !1, Module.prototype.require = og;
    }
  };
}
const packageJson = require("react-native-web/package.json"), IGNORES = process.env.TAMAGUI_IGNORE_BUNDLE_ERRORS;
var _a;
const extraIgnores = IGNORES === "true" ? [] : (_a = process.env.TAMAGUI_IGNORE_BUNDLE_ERRORS) == null ? void 0 : _a.split(","), knownIgnorableModules = {
  "expo-modules": !0,
  solito: !0,
  ...Object.fromEntries((extraIgnores == null ? void 0 : extraIgnores.map((k) => [k, !0])) || [])
}, hasWarnedForModules = /* @__PURE__ */ new Set(), allowedIgnores = {
  "expo-constants": !0
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getNameToPaths,
  registerRequire,
  setRequireResult
});
