// @tamagui/static/dist/esm/extractor/extractToClassNames.js
import * as path from "path";
import { basename } from "path";
import * as util from "util";
import generate from "@babel/generator";
import * as t from "@babel/types";
import { concatClassName } from "@tamagui/helpers";
import { requireTamaguiCore } from "../helpers/requireTamaguiCore";
import { babelParse } from "./babelParse";
import { buildClassName } from "./buildClassName";
import { ensureImportingConcat } from "./ensureImportingConcat";
import { isSimpleSpread } from "./extractHelpers";
import { extractMediaStyle } from "./extractMediaStyle";
import { getPrefixLogs } from "./getPrefixLogs";
import { hoistClassNames } from "./hoistClassNames";
import { logLines } from "./logLines";
import { getFontFamilyClassNameFromProps } from "./propsToFontFamilyCache";
import { timer } from "./timer";
const mergeStyleGroups = {
  shadowOpacity: !0,
  shadowRadius: !0,
  shadowColor: !0,
  shadowOffset: !0
};
async function extractToClassNames({
  extractor,
  source,
  sourcePath,
  options,
  shouldPrintDebug
}) {
  const tm = timer(), { getStylesAtomic } = requireTamaguiCore("web");
  if (sourcePath?.includes("node_modules"))
    return null;
  if (shouldPrintDebug && console.warn(`--- ${sourcePath} --- 

`), typeof source != "string")
    throw new Error("`source` must be a string of javascript");
  if (sourcePath && !path.isAbsolute(sourcePath))
    throw new Error("`sourcePath` must be an absolute path to a .js file");
  /.[tj]sx?$/i.test(sourcePath || "") || console.warn(`${sourcePath?.slice(0, 100)} - bad filename.`), !options.disableExtraction && !options._disableLoadTamagui && await extractor.loadTamagui(options);
  const shouldLogTiming = options.logTimings ?? !0, start = Date.now(), mem = process.env.TAMAGUI_SHOW_MEMORY_USAGE && shouldLogTiming ? process.memoryUsage() : null;
  let ast;
  try {
    ast = babelParse(source, sourcePath);
  } catch (err) {
    throw console.error("babel parse error:", sourcePath?.slice(0, 100)), err;
  }
  tm.mark("babel-parse", shouldPrintDebug === "verbose");
  const cssMap = /* @__PURE__ */ new Map(), existingHoists = {};
  let hasFlattened = !1;
  const res = await extractor.parse(ast, {
    shouldPrintDebug,
    ...options,
    sourcePath,
    extractStyledDefinitions: !0,
    onStyleRule(identifier, rules) {
      const css = rules.join(";");
      shouldPrintDebug && console.log(`adding styled() rule: .${identifier} ${css}`), cssMap.set(`.${identifier}`, { css, commentTexts: [] });
    },
    getFlattenedNode: ({ tag }) => (hasFlattened = !0, tag),
    onExtractTag: ({
      parserProps,
      attrs,
      node,
      attemptEval,
      jsxPath,
      originalNodeName,
      filePath,
      lineNumbers,
      programPath,
      isFlattened,
      config,
      completeProps,
      staticConfig
    }) => {
      if (staticConfig.acceptsClassName === !1) {
        shouldPrintDebug && console.log("bail, acceptsClassName is false");
        return;
      }
      const didFlattenThisTag = hasFlattened;
      hasFlattened = !1;
      let finalClassNames = [];
      const finalAttrs = [];
      let finalStyles = [], viewStyles = {};
      for (const attr of attrs)
        attr.type === "style" && (viewStyles = {
          ...viewStyles,
          ...attr.value
        });
      const ensureNeededPrevStyle = (style) => {
        if (!Object.keys(style).some((key) => mergeStyleGroups[key]))
          return style;
        for (const k in mergeStyleGroups)
          k in viewStyles && (style[k] = style[k] ?? viewStyles[k]);
        return style;
      }, addStyles = (style) => {
        if (!style)
          return [];
        const styleWithPrev = ensureNeededPrevStyle(style), res2 = getStylesAtomic(styleWithPrev);
        return res2.length && (finalStyles = [...finalStyles, ...res2]), res2;
      };
      let lastMediaImportance = 1;
      for (const attr of attrs)
        switch (attr.type) {
          case "style": {
            if (isFlattened) {
              const styles2 = addStyles(attr.value), newFontFamily = getFontFamilyClassNameFromProps(attr.value) || "", newClassNames = concatClassName(
                styles2.map((x) => x.identifier).join(" ") + newFontFamily
              ), existing = finalClassNames.find(
                (x) => x.type == "StringLiteral"
              );
              if (existing) {
                let previous = existing.value;
                newFontFamily && (shouldPrintDebug && console.log(` newFontFamily: ${newFontFamily}`), previous = previous.replace(/font_[a-z]+/i, "")), existing.value = `${previous} ${newClassNames}`;
              } else
                finalClassNames = [...finalClassNames, t.stringLiteral(newClassNames)];
            } else {
              const styles2 = getStylesAtomic(attr.value);
              finalStyles = [...finalStyles, ...styles2];
              for (const style of styles2) {
                const prop = style.pseudo ? `${style.property}-${style.pseudo}` : style.property;
                finalAttrs.push(
                  t.jsxAttribute(t.jsxIdentifier(prop), t.stringLiteral(style.identifier))
                );
              }
            }
            break;
          }
          case "attr": {
            const val = attr.value;
            if (t.isJSXSpreadAttribute(val))
              isSimpleSpread(val) && finalClassNames.push(
                t.logicalExpression(
                  "&&",
                  val.argument,
                  t.memberExpression(val.argument, t.identifier("className"))
                )
              );
            else if (val.name.name === "className") {
              const value = val.value;
              if (value)
                try {
                  const evaluatedValue = attemptEval(value);
                  finalClassNames.push(t.stringLiteral(evaluatedValue));
                } catch {
                  finalClassNames.push(value.expression);
                }
              continue;
            }
            finalAttrs.push(val);
            break;
          }
          case "ternary": {
            const mediaExtraction = extractMediaStyle(
              parserProps,
              attr.value,
              jsxPath,
              extractor.getTamagui(),
              sourcePath || "",
              lastMediaImportance,
              shouldPrintDebug
            );
            if (shouldPrintDebug && mediaExtraction && console.log(
              "ternary (mediaStyles)",
              mediaExtraction.ternaryWithoutMedia?.inlineMediaQuery ?? "",
              mediaExtraction.mediaStyles.map((x) => x.identifier).join(".")
            ), !mediaExtraction) {
              addTernaryStyle(
                attr.value,
                addStyles(attr.value.consequent),
                addStyles(attr.value.alternate)
              );
              continue;
            }
            lastMediaImportance++, mediaExtraction.mediaStyles && (finalStyles = [...finalStyles, ...mediaExtraction.mediaStyles]), mediaExtraction.ternaryWithoutMedia ? addTernaryStyle(
              mediaExtraction.ternaryWithoutMedia,
              mediaExtraction.mediaStyles,
              []
            ) : finalClassNames = [
              ...finalClassNames,
              ...mediaExtraction.mediaStyles.map((x) => t.stringLiteral(x.identifier))
            ];
            break;
          }
        }
      function addTernaryStyle(ternary, a, b) {
        const cCN = a.map((x) => x.identifier).join(" "), aCN = b.map((x) => x.identifier).join(" ");
        a.length && b.length ? finalClassNames.push(
          t.conditionalExpression(
            ternary.test,
            t.stringLiteral(cCN),
            t.stringLiteral(aCN)
          )
        ) : finalClassNames.push(
          t.conditionalExpression(
            ternary.test,
            t.stringLiteral(" " + cCN),
            t.stringLiteral(" " + aCN)
          )
        );
      }
      if (shouldPrintDebug && console.log(
        `  finalClassNames
`,
        logLines(finalClassNames.map((x) => x.value).join(" "))
      ), node.attributes = finalAttrs, finalClassNames.length) {
        const extraClassNames = (() => {
          let value = "";
          return isFlattened && (process.env.TAMAGUI_DEBUG_OPTIMIZATIONS && (value += "is_tamagui_flattened"), staticConfig.componentName && (value += ` is_${staticConfig.componentName}`)), value;
        })(), names = buildClassName(finalClassNames, extraClassNames), nameExpr = names ? hoistClassNames(jsxPath, existingHoists, names) : null;
        let expr = nameExpr;
        if (nameExpr && !t.isIdentifier(nameExpr) && didFlattenThisTag) {
          ensureImportingConcat(programPath);
          const simpleSpreads = attrs.filter(
            (x) => t.isJSXSpreadAttribute(x.value) && isSimpleSpread(x.value)
          );
          expr = t.callExpression(t.identifier("concatClassName"), [
            expr,
            ...simpleSpreads.map((val) => val.value.argument)
          ]);
        }
        node.attributes.push(
          t.jsxAttribute(t.jsxIdentifier("className"), t.jsxExpressionContainer(expr))
        );
      }
      const comment = util.format(
        "/* %s:%s (%s) */",
        filePath,
        lineNumbers,
        originalNodeName
      );
      for (const { identifier, rules } of finalStyles) {
        const className = `.${identifier}`;
        if (cssMap.has(className)) {
          if (comment) {
            const val = cssMap.get(className);
            val.commentTexts.push(comment), cssMap.set(className, val);
          }
        } else
          rules.length && cssMap.set(className, {
            css: rules.join(`
`),
            commentTexts: [comment]
          });
      }
    }
  });
  if (!res || !res.modified && !res.optimized && !res.flattened && !res.styled)
    return shouldPrintDebug && console.log("no res or none modified", res), null;
  const styles = Array.from(cssMap.values()).map((x) => x.css).join(`
`).trim(), result = generate(
    ast,
    {
      concise: !1,
      filename: sourcePath,
      // this makes the debug output terrible, and i think sourcemap works already
      retainLines: !1,
      sourceFileName: sourcePath,
      sourceMaps: !0
    },
    source
  );
  if (shouldPrintDebug && (console.log(
    `
 -------- output code ------- 

`,
    result.code.split(`
`).filter((x) => !x.startsWith("//")).join(`
`)
  ), console.log(`
 -------- output style -------- 

`, styles)), shouldLogTiming) {
    const memUsed = mem ? Math.round((process.memoryUsage().heapUsed - mem.heapUsed) / 1024 / 1204 * 10) / 10 : 0, path2 = basename(sourcePath || "").replace(/\.[jt]sx?$/, "").slice(0, 22).trim().padStart(24), numOptimized = `${res.optimized + res.styled}`.padStart(3), numFound = `${res.found + res.styled}`.padStart(3), numFlattened = `${res.flattened}`.padStart(3), memory = memUsed ? ` ${memUsed}MB` : "", timingStr = `${Date.now() - start}ms`.padStart(6), pre = getPrefixLogs(options), memStr = memory ? `(${memory})` : "";
    console.log(
      `${pre} ${path2}   \xB7  ${numFound} found   \xB7  ${numOptimized} opt   \xB7  ${numFlattened} flat  ${timingStr} ${memStr}`
    );
  }
  return {
    ast,
    styles,
    js: result.code,
    map: result.map
  };
}
export {
  extractToClassNames
};
