// @tamagui/static/dist/esm/extractor/extractMediaStyle.native.js
import * as t from "@babel/types";
import { MEDIA_SEP } from "../constants";
import { requireTamaguiCore } from "../helpers/requireTamaguiCore";
import { isPresent, isValidImport } from "./extractHelpers";
function extractMediaStyle(props, ternary, jsxPath, tamaguiConfig, sourcePath, importance = 0, shouldPrintDebug = !1) {
  const { getStylesAtomic, mediaObjectToString } = requireTamaguiCore("web"), mt = getMediaQueryTernary(props, ternary, jsxPath, sourcePath);
  if (!mt)
    return null;
  const { key } = mt;
  if (!tamaguiConfig.media[key])
    return console.error(`Media query "${key}" not found: ${Object.keys(tamaguiConfig.media)}`), null;
  const getStyleObj = (styleObj, negate = !1) => styleObj ? { styleObj, negate } : null, styleOpts = [
    getStyleObj(ternary.consequent, !1),
    getStyleObj(ternary.alternate, !0)
  ].filter(isPresent);
  if (shouldPrintDebug && !styleOpts.length)
    return console.log("  media query, no styles?"), null;
  const mediaKeyPrecendence = Object.keys(tamaguiConfig.media).reduce((acc, cur, i) => (acc[cur] = new Array(importance + 1).fill(":root").join(""), acc), {});
  let mediaStyles = [];
  for (const { styleObj, negate } of styleOpts) {
    const singleMediaStyles = getStylesAtomic(styleObj).map((style) => {
      const negKey = negate ? "0" : "", ogPrefix = style.identifier.slice(0, style.identifier.indexOf("-") + 1), identifier = `${style.identifier.replace(
        ogPrefix,
        `${ogPrefix}${MEDIA_SEP}${key}${negKey}${MEDIA_SEP}`
      )}`, className = `.${identifier}`, mediaSelector = mediaObjectToString(tamaguiConfig.media[key]), mediaQuery = `${negate ? "not all" : "screen"} and ${mediaSelector}`, precendencePrefix = mediaKeyPrecendence[key], styleInner = style.rules.map((rule) => rule.replace(style.identifier, identifier)).join(";");
      let styleRule = "";
      return styleInner.includes("@media") ? styleRule = styleInner.replace("{", ` and ${mediaQuery} {`) : styleRule = `@media ${mediaQuery} { ${precendencePrefix}${styleInner} }`, {
        ...style,
        identifier,
        className,
        rules: [styleRule]
      };
    });
    shouldPrintDebug === "verbose" && console.log("  media styles:", importance, styleObj, singleMediaStyles.map((x) => x.identifier).join(", ")), mediaStyles = [...mediaStyles, ...singleMediaStyles];
  }
  return ternary.remove(), { mediaStyles, ternaryWithoutMedia: mt.ternaryWithoutMedia };
}
function getMediaQueryTernary(props, ternary, jsxPath, sourcePath) {
  if (t.isLogicalExpression(ternary.test) && ternary.test.operator === "&&") {
    const mediaLeft = getMediaInfoFromExpression(
      props,
      ternary.test.left,
      jsxPath,
      sourcePath,
      ternary.inlineMediaQuery
    );
    if (mediaLeft)
      return {
        ...mediaLeft,
        ternaryWithoutMedia: {
          ...ternary,
          test: ternary.test.right
        }
      };
  }
  const result = getMediaInfoFromExpression(
    props,
    ternary.test,
    jsxPath,
    sourcePath,
    ternary.inlineMediaQuery
  );
  return result ? {
    ...result,
    ternaryWithoutMedia: null
  } : null;
}
function getMediaInfoFromExpression(props, test, jsxPath, sourcePath, inlineMediaQuery) {
  if (inlineMediaQuery)
    return {
      key: inlineMediaQuery,
      bindingName: inlineMediaQuery
    };
  if (t.isMemberExpression(test) && t.isIdentifier(test.object) && t.isIdentifier(test.property)) {
    const name = test.object.name, key = test.property.name, binding = jsxPath.scope.getAllBindings()[name];
    if (!binding)
      return !1;
    const bindingNode = binding.path?.node;
    return !t.isVariableDeclarator(bindingNode) || !bindingNode.init || !isValidMediaCall(props, jsxPath, bindingNode.init, sourcePath) ? !1 : { key, bindingName: name };
  }
  if (t.isIdentifier(test)) {
    const key = test.name, node = jsxPath.scope.getBinding(test.name)?.path?.node;
    return !t.isVariableDeclarator(node) || !node.init || !isValidMediaCall(props, jsxPath, node.init, sourcePath) ? !1 : { key, bindingName: key };
  }
  return null;
}
function isValidMediaCall(props, jsxPath, init, sourcePath) {
  if (!init || !t.isCallExpression(init) || !t.isIdentifier(init.callee) || init.callee.name !== "useMedia")
    return !1;
  const mediaBinding = jsxPath.scope.getAllBindings().useMedia;
  if (!mediaBinding)
    return !1;
  const useMediaImport = mediaBinding.path.parent;
  return !(!t.isImportDeclaration(useMediaImport) || !isValidImport(props, sourcePath));
}
export {
  extractMediaStyle,
  isValidMediaCall
};
