// react-native-web-internals/dist/cjs/index.native.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf, __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
}, __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default")), __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: !0 }) : target,
  mod
)), __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var src_exports = {};
__export(src_exports, {
  AccessibilityUtil: () => import_AccessibilityUtil.default,
  ImageLoader: () => import_ImageLoader.default,
  InteractionManager: () => import_InteractionManager.InteractionManager,
  Platform: () => import_Platform.default,
  StyleSheet: () => import_StyleSheet2.default,
  TextAncestorContext: () => import_TextAncestorContext.TextAncestorContext,
  TextInputState: () => import_TextInputState.default,
  UIManager: () => import_UIManager.default,
  canUseDOM: () => import_canUseDOM.default,
  colorProps: () => import_colorProps.colorProps,
  createDOMProps: () => import_createDOMProps.default,
  createEventHandle: () => import_createEventHandle.default,
  createSheet: () => import_dom.createSheet,
  dismissKeyboard: () => import_dismissKeyboard.default,
  flattenStyle: () => import_StyleSheet.flatten,
  getBoundingClientRect: () => import_getBoundingClientRect.default,
  isSelectionValid: () => import_isSelectionValid.default,
  isWebColor: () => import_isWebColor.default,
  multiplyStyleLengthValue: () => import_multiplyStyleLengthValue.default,
  normalizeColor: () => import_normalizeColor.default,
  pick: () => import_pick.default,
  processColor: () => import_processColor.processColor,
  requestIdleCallback: () => import_requestIdleCallback.default,
  setValueForStyles: () => import_setValueForStyles.default,
  unitlessNumbers: () => import_unitlessNumbers.default,
  useElementLayout: () => import_useElementLayout.default,
  useEvent: () => import_useEvent.default,
  useHover: () => import_useHover.default,
  useLayoutEffect: () => import_useLayoutEffect.default,
  useMergeRefs: () => import_useMergeRefs.useMergeRefs,
  usePlatformMethods: () => import_usePlatformMethods.usePlatformMethods,
  useStable: () => import_useStable.default
});
module.exports = __toCommonJS(src_exports);
__reExport(src_exports, require("./modules/AssetRegistry/index"), module.exports);
__reExport(src_exports, require("./modules/forwardedProps/index"), module.exports);
__reExport(src_exports, require("./modules/mergeRefs/index"), module.exports);
__reExport(src_exports, require("./modules/modality/index"), module.exports);
__reExport(src_exports, require("./modules/useLocale/index"), module.exports);
var import_usePlatformMethods = require("./modules/usePlatformMethods/index"), import_TextAncestorContext = require("./TextAncestorContext");
__reExport(src_exports, require("@tamagui/react-native-use-pressable"), module.exports);
__reExport(src_exports, require("@tamagui/react-native-use-responder-events"), module.exports);
var import_colorProps = require("./colorProps"), import_AccessibilityUtil = __toESM(require("./modules/AccessibilityUtil/index")), import_canUseDOM = __toESM(require("./modules/canUseDOM")), import_createDOMProps = __toESM(require("./modules/createDOMProps/index")), import_createEventHandle = __toESM(require("./modules/createEventHandle/index")), import_dismissKeyboard = __toESM(require("./modules/dismissKeyboard/index")), import_getBoundingClientRect = __toESM(require("./modules/getBoundingClientRect/index")), import_ImageLoader = __toESM(require("./modules/ImageLoader/index")), import_isSelectionValid = __toESM(require("./modules/isSelectionValid/index")), import_isWebColor = __toESM(require("./modules/isWebColor/index")), import_multiplyStyleLengthValue = __toESM(require("./modules/multiplyStyleLengthValue/index")), import_normalizeColor = __toESM(require("./modules/normalizeColor/index")), import_pick = __toESM(require("./modules/pick/index")), import_Platform = __toESM(require("./modules/Platform/index"));
__reExport(src_exports, require("./StyleSheet/preprocess"), module.exports);
var import_StyleSheet = require("./StyleSheet/index"), import_dom = require("./StyleSheet/dom/index"), import_requestIdleCallback = __toESM(require("./modules/requestIdleCallback/index")), import_setValueForStyles = __toESM(require("./modules/setValueForStyles/index")), import_TextInputState = __toESM(require("./modules/TextInputState/index")), import_UIManager = __toESM(require("./modules/UIManager/index")), import_unitlessNumbers = __toESM(require("./modules/unitlessNumbers/index")), import_useElementLayout = __toESM(require("./modules/useElementLayout/index")), import_useEvent = __toESM(require("./modules/useEvent/index")), import_useHover = __toESM(require("./modules/useHover/index")), import_useLayoutEffect = __toESM(require("./modules/useLayoutEffect/index")), import_useStable = __toESM(require("./modules/useStable/index")), import_InteractionManager = require("./modules/InteractionManager");
__reExport(src_exports, require("./modules/invariant"), module.exports);
var import_processColor = require("./modules/processColor/index"), import_StyleSheet2 = __toESM(require("./StyleSheet/index")), import_useMergeRefs = require("./modules/useMergeRefs/index");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AccessibilityUtil,
  ImageLoader,
  InteractionManager,
  Platform,
  StyleSheet,
  TextAncestorContext,
  TextInputState,
  UIManager,
  canUseDOM,
  colorProps,
  createDOMProps,
  createEventHandle,
  createSheet,
  dismissKeyboard,
  flattenStyle,
  getBoundingClientRect,
  isSelectionValid,
  isWebColor,
  multiplyStyleLengthValue,
  normalizeColor,
  pick,
  processColor,
  requestIdleCallback,
  setValueForStyles,
  unitlessNumbers,
  useElementLayout,
  useEvent,
  useHover,
  useLayoutEffect,
  useMergeRefs,
  usePlatformMethods,
  useStable,
  ...require("./modules/AssetRegistry/index"),
  ...require("./modules/forwardedProps/index"),
  ...require("./modules/mergeRefs/index"),
  ...require("./modules/modality/index"),
  ...require("./modules/useLocale/index"),
  ...require("@tamagui/react-native-use-pressable"),
  ...require("@tamagui/react-native-use-responder-events"),
  ...require("./StyleSheet/preprocess"),
  ...require("./modules/invariant")
});
