// tamagui/dist/esm/views/Grid.js
import { isWeb } from "@tamagui/core";
import { XStack } from "@tamagui/stacks";
import React from "react";
import { jsx } from "react/jsx-runtime";
function Grid({ children, columns, itemMinWidth = 200, gap }) {
  if (isWeb)
    return /* @__PURE__ */ jsx(
      "div",
      {
        style: {
          gap,
          display: "grid",
          justifyContent: "stretch",
          // gridTemplateRows: 'repeat(4, 1fr)',
          gridTemplateColumns: `repeat( auto-fit, minmax(${itemMinWidth}px, 1fr) )`
          // gridTemplateColumns: '1fr 1fr',
        },
        children
      }
    );
  const childrenList = React.Children.toArray(children);
  return /* @__PURE__ */ jsx(XStack, { alignItems: "center", justifyContent: "center", flexWrap: "wrap", children: childrenList.map((child, i) => child ? /* @__PURE__ */ jsx(
    XStack,
    {
      flex: 1,
      minWidth: itemMinWidth,
      marginRight: gap,
      marginBottom: gap,
      children: child
    },
    i
  ) : null) });
}
export {
  Grid
};
