// @tamagui/static/dist/cjs/extractor/loadTamagui.native.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf, __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: !0 }) : target,
  mod
)), __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var loadTamagui_exports = {};
__export(loadTamagui_exports, {
  TamaguiProjectInfo: () => import_bundleConfig.TamaguiProjectInfo,
  getOptions: () => getOptions,
  loadTamagui: () => loadTamagui,
  loadTamaguiSync: () => loadTamaguiSync,
  resolveWebOrNativeSpecificEntry: () => resolveWebOrNativeSpecificEntry,
  watchTamaguiConfig: () => watchTamaguiConfig
});
module.exports = __toCommonJS(loadTamagui_exports);
var import_path = require("path"), import_cli_color = require("@tamagui/cli-color"), import_config_default = require("@tamagui/config-default"), import_esbuild = __toESM(require("esbuild")), import_fs_extra = require("fs-extra"), import_constants = require("../constants"), import_requireTamaguiCore = require("../helpers/requireTamaguiCore"), import_registerRequire = require("../registerRequire"), import_bundleConfig = require("./bundleConfig"), import_generateTamaguiStudioConfig = require("./generateTamaguiStudioConfig"), import_getTamaguiConfigPathFromOptionsConfig = require("./getTamaguiConfigPathFromOptionsConfig");
const getFilledOptions = (propsIn) => ({
  // defaults
  platform: "native",
  config: "tamagui.config.ts",
  components: ["tamagui"],
  ...propsIn
});
async function loadTamagui(propsIn) {
  const props = getFilledOptions(propsIn);
  await generateThemesAndLog(props);
  const bundleInfo = await (0, import_bundleConfig.getBundledConfig)(props);
  if (!bundleInfo)
    return console.warn(
      "No bundled config generated, maybe an error in bundling. Set DEBUG=tamagui and re-run to get logs."
    ), null;
  if (!(0, import_bundleConfig.hasBundledConfigChanged)())
    return bundleInfo;
  if (bundleInfo) {
    const { createTamagui } = (0, import_requireTamaguiCore.requireTamaguiCore)(props.platform), config = createTamagui(bundleInfo.tamaguiConfig);
    if (props.outputCSS) {
      (0, import_cli_color.colorLog)(import_cli_color.Color.FgYellow, `    \u27A1 [tamagui] outputCSS: ${props.outputCSS}
`);
      const css = config.getCSS();
      await (0, import_fs_extra.writeFile)(props.outputCSS, css);
    }
  }
  return process.env.NODE_ENV === "development" && await (0, import_generateTamaguiStudioConfig.generateTamaguiStudioConfig)(props, bundleInfo), bundleInfo;
}
let waiting = !1, hasLoggedOnce = !1;
const generateThemesAndLog = async (options) => {
  if (!waiting && options.themeBuilder)
    try {
      waiting = !0, await new Promise((res) => setTimeout(res, 30));
      const didGenerate = await (0, import_generateTamaguiStudioConfig.generateTamaguiThemes)(options);
      if (!hasLoggedOnce || didGenerate) {
        hasLoggedOnce = !0;
        const whitespaceBefore = "    ";
        (0, import_cli_color.colorLog)(import_cli_color.Color.FgYellow, `${whitespaceBefore}\u27A1 [tamagui] Generated themes:`), (0, import_cli_color.colorLog)(
          import_cli_color.Color.Dim,
          `
${whitespaceBefore}${(0, import_path.relative)(process.cwd(), options.themeBuilder.output)}`
        );
      }
    } finally {
      waiting = !1;
    }
}, last = {}, lastVersion = {};
function loadTamaguiSync({
  forceExports,
  cacheKey,
  ...propsIn
}) {
  const key = JSON.stringify(propsIn);
  if (last[key] && !(0, import_bundleConfig.hasBundledConfigChanged)() && (!lastVersion[key] || lastVersion[key] === cacheKey))
    return last[key];
  lastVersion[key] = cacheKey || "";
  const props = getFilledOptions(propsIn);
  process.env.IS_STATIC = "is_static", process.env.TAMAGUI_IS_SERVER = "true";
  const { unregister } = (0, import_registerRequire.registerRequire)(props.platform, {
    proxyWormImports: !!forceExports
  });
  try {
    const devValueOG = globalThis.__DEV__;
    globalThis.__DEV__ = process.env.NODE_ENV === "development";
    try {
      let tamaguiConfig = null;
      if (propsIn.config) {
        const configPath = (0, import_getTamaguiConfigPathFromOptionsConfig.getTamaguiConfigPathFromOptionsConfig)(propsIn.config), exp = require(configPath);
        if (tamaguiConfig = exp.default || exp, !tamaguiConfig || !tamaguiConfig.parsed) {
          const confPath = require.resolve(configPath);
          throw new Error(`Can't find valid config in ${confPath}:
          
  Be sure you "export default" the config.`);
        }
        if (tamaguiConfig) {
          const { createTamagui } = (0, import_requireTamaguiCore.requireTamaguiCore)(props.platform);
          createTamagui(tamaguiConfig);
        }
      }
      const components = (0, import_bundleConfig.loadComponents)(props, forceExports);
      if (!components)
        throw new Error("No components loaded");
      process.env.DEBUG === "tamagui" && console.log("components", components), process.env.IS_STATIC = void 0, globalThis.__DEV__ = devValueOG;
      const info = {
        components,
        tamaguiConfig,
        nameToPaths: (0, import_registerRequire.getNameToPaths)()
      };
      return propsIn.config && (0, import_generateTamaguiStudioConfig.generateTamaguiStudioConfigSync)(props, info), last[key] = {
        ...info,
        cached: !0
      }, info;
    } catch (err) {
      err instanceof Error ? !import_constants.SHOULD_DEBUG && !forceExports ? (console.warn(
        "Error loading tamagui.config.ts (set DEBUG=tamagui to see full stack), running tamagui without custom config"
      ), console.log(`

    ${err.message}

`)) : import_constants.SHOULD_DEBUG && console.error(err) : console.error("Error loading tamagui.config.ts", err);
      const { createTamagui } = (0, import_requireTamaguiCore.requireTamaguiCore)(props.platform);
      return {
        components: [],
        tamaguiConfig: createTamagui((0, import_config_default.getDefaultTamaguiConfig)()),
        nameToPaths: {}
      };
    }
  } finally {
    unregister();
  }
}
async function getOptions({
  root = process.cwd(),
  tsconfigPath = "tsconfig.json",
  tamaguiOptions,
  host,
  debug
} = {}) {
  const dotDir = (0, import_path.join)(root, ".tamagui");
  let pkgJson = {};
  try {
    pkgJson = await (0, import_fs_extra.readJSON)((0, import_path.join)(root, "package.json"));
  } catch {
  }
  return {
    mode: process.env.NODE_ENV === "production" ? "production" : "development",
    root,
    host: host || "127.0.0.1",
    pkgJson,
    debug,
    tsconfigPath,
    tamaguiOptions: {
      platform: "native",
      components: ["tamagui"],
      ...tamaguiOptions,
      config: await getDefaultTamaguiConfigPath(root, tamaguiOptions == null ? void 0 : tamaguiOptions.config)
    },
    paths: {
      dotDir,
      conf: (0, import_path.join)(dotDir, "tamagui.config.json"),
      types: (0, import_path.join)(dotDir, "types.json")
    }
  };
}
function resolveWebOrNativeSpecificEntry(entry) {
  const workspaceRoot = (0, import_path.resolve)(), resolved = require.resolve(entry, { paths: [workspaceRoot] }), ext = (0, import_path.extname)(resolved), fileName = (0, import_path.basename)(resolved).replace(ext, ""), specificExt = "native", specificFile = (0, import_path.join)((0, import_path.dirname)(resolved), fileName + "." + specificExt + ext);
  return (0, import_fs_extra.existsSync)(specificFile) ? specificFile : entry;
}
const defaultPaths = ["tamagui.config.ts", (0, import_path.join)("src", "tamagui.config.ts")];
let hasWarnedOnce = !1;
async function getDefaultTamaguiConfigPath(root, configPath) {
  const searchPaths = [
    ...new Set(
      [configPath, ...defaultPaths].filter(Boolean).map((p) => (0, import_path.join)(root, p))
    )
  ];
  for (const path of searchPaths)
    if (await (0, import_fs_extra.pathExists)(path))
      return path;
  hasWarnedOnce || (hasWarnedOnce = !0, console.warn(`Warning: couldn't find tamagui.config.ts in the following paths given configuration "${configPath}":
    ${searchPaths.join(`
  `)}
  `));
}
async function watchTamaguiConfig(tamaguiOptions) {
  var _a;
  const options = await getOptions({ tamaguiOptions });
  if (!options.tamaguiOptions.config)
    throw new Error("No config");
  if (process.env.NODE_ENV === "production")
    return {
      dispose() {
      }
    };
  const disposeConfigWatcher = await esbuildWatchFiles(
    options.tamaguiOptions.config,
    () => {
      (0, import_generateTamaguiStudioConfig.generateTamaguiStudioConfig)(options.tamaguiOptions, null, !0);
    }
  ), themeBuilderInput = (_a = options.tamaguiOptions.themeBuilder) == null ? void 0 : _a.input;
  let disposeThemesWatcher;
  if (themeBuilderInput) {
    let inputPath = themeBuilderInput;
    try {
      inputPath = require.resolve(themeBuilderInput);
    } catch {
    }
    disposeThemesWatcher = await esbuildWatchFiles(inputPath, () => {
      generateThemesAndLog(options.tamaguiOptions);
    });
  }
  return {
    dispose() {
      disposeConfigWatcher(), disposeThemesWatcher == null || disposeThemesWatcher();
    }
  };
}
async function esbuildWatchFiles(entry, onChanged) {
  let hasRunOnce = !1;
  const context = await import_esbuild.default.context({
    bundle: !0,
    entryPoints: [entry],
    resolveExtensions: [".ts", ".tsx", ".js", ".jsx", ".mjs"],
    logLevel: "silent",
    write: !1,
    plugins: [
      {
        name: "on-rebuild",
        setup({ onEnd }) {
          onEnd(() => {
            hasRunOnce ? onChanged() : hasRunOnce = !0;
          });
        }
      }
    ]
  });
  return context.watch(), () => {
    context.dispose();
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TamaguiProjectInfo,
  getOptions,
  loadTamagui,
  loadTamaguiSync,
  resolveWebOrNativeSpecificEntry,
  watchTamaguiConfig
});
