// @tamagui/static/dist/cjs/extractor/validHTMLAttributes.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var validHTMLAttributes_exports = {};
__export(validHTMLAttributes_exports, {
  validHTMLAttributes: () => validHTMLAttributes
});
module.exports = __toCommonJS(validHTMLAttributes_exports);
const validHTMLAttributes = {
  autocomplete: !0,
  border: !0,
  contenteditable: !0,
  crossorigin: !0,
  dir: !0,
  draggable: !0,
  enctype: !0,
  formenctype: !0,
  formmethod: !0,
  formtarget: !0,
  inputmode: !0,
  kind: !0,
  link: !0,
  method: !0,
  preload: !0,
  referrerpolicy: !0,
  rel: !0,
  rev: !0,
  role: !0,
  sandbox: !0,
  shape: !0,
  spellcheck: !0,
  target: !0,
  translate: !0,
  type: !0,
  wrap: !0,
  "aria-autocomplete": !0,
  "aria-busy": !0,
  "aria-checked": !0,
  "aria-current": !0,
  "aria-disabled": !0,
  "aria-expanded": !0,
  "aria-haspopup": !0,
  "aria-hidden": !0,
  "aria-invalid": !0,
  "aria-polite": !0,
  "aria-modal": !0,
  "aria-multiline": !0,
  "aria-multiselectable": !0,
  "aria-orientation": !0,
  "aria-pressed": !0,
  "aria-readonly": !0,
  "aria-relevant": !0,
  "aria-required": !0,
  "aria-selected": !0,
  "aria-sort": !0
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  validHTMLAttributes
});
