// react-native-web-lite/dist/esm/index.native.js
import { default as default2 } from "./createElement/index";
import {
  processColor,
  Platform,
  UIManager,
  canUseDOM,
  dismissKeyboard,
  isWebColor,
  mergeRefs,
  useMergeRefs,
  useEvent,
  useHover,
  ImageLoader
} from "react-native-web-internals";
import { default as default3 } from "./render/index";
import { unmountComponentAtNode } from "react-dom";
import { default as default4 } from "./NativeModules/index";
import { findNodeHandle } from "./findNodeHandle";
import { default as default5 } from "./vendor/react-native/Animated/Easing";
import { default as default6 } from "./vendor/react-native/Animated/Animated";
import { default as default7 } from "./vendor/react-native/NativeEventEmitter/index";
import { default as default8 } from "./AccessibilityInfo/index";
import { default as default9 } from "./Alert/index";
import { default as default10 } from "./Appearance/index";
import { default as default11 } from "./AppRegistry/index";
import { default as default12 } from "./AppState/index";
import { default as default13 } from "./BackHandler/index";
import { default as default14 } from "./Clipboard/index";
import { default as default15 } from "./DeviceInfo/index";
import { default as default16 } from "./DeviceEmitter";
import { default as default17 } from "./DeviceEmitter";
import { default as default18 } from "./Dimensions/index";
import { default as default19 } from "./I18nManager/index";
import { default as default20 } from "./Keyboard/index";
import { default as default21 } from "./Linking/index";
import { default as default22 } from "./PanResponder/index";
import { default as default23 } from "./PixelRatio/index";
import { default as default24 } from "./Share/index";
import { default as default25 } from "./Vibration/index";
import { default as default26 } from "./UnimplementedView";
import { default as default27 } from "./UnimplementedView";
import { default as default28 } from "./UnimplementedView";
import { default as default29 } from "./UnimplementedView";
import { default as default30 } from "./UnimplementedView";
import { default as default31 } from "./UnimplementedView";
import { default as default32 } from "./UnimplementedView";
import { default as default33 } from "./TouchableOpacity";
import { default as default34 } from "./TouchableWithoutFeedback";
import { default as default35 } from "./ActivityIndicator/index";
import { default as default36 } from "./Image/index";
import { default as default37 } from "./ImageBackground/index";
import { default as default38 } from "./KeyboardAvoidingView/index";
import { default as default39 } from "./Modal/index";
import { default as default40 } from "./Pressable/index";
import { default as default41 } from "./RefreshControl/index";
import { default as default42 } from "./SafeAreaView/index";
import { default as default43 } from "./ScrollView/index";
import { default as default44 } from "./StatusBar/index";
import { default as default45 } from "./Text/index";
import { default as default46 } from "./TextInput/index";
import { default as default47 } from "./View/index";
import { default as default48 } from "./LogBox/index";
import { default as default49 } from "./useColorScheme/index";
import { useLocaleContext } from "./useLocaleContext/index";
import { default as default50 } from "./useWindowDimensions/index";
export * from "react-native-web-internals";
function requireNativeComponent(name) {
  return function() {
    return null;
  };
}
export {
  default8 as AccessibilityInfo,
  default35 as ActivityIndicator,
  default9 as Alert,
  default6 as Animated,
  default11 as AppRegistry,
  default12 as AppState,
  default10 as Appearance,
  default13 as BackHandler,
  default14 as Clipboard,
  default16 as DeviceEmitter,
  default17 as DeviceEventEmitter,
  default15 as DeviceInfo,
  default18 as Dimensions,
  default26 as DrawerLayoutAndroid,
  default5 as Easing,
  default29 as FlatList,
  default19 as I18nManager,
  default36 as Image,
  default37 as ImageBackground,
  ImageLoader,
  default20 as Keyboard,
  default38 as KeyboardAvoidingView,
  default21 as Linking,
  default48 as LogBox,
  default39 as Modal,
  default7 as NativeEventEmitter,
  default4 as NativeModules,
  default22 as PanResponder,
  default23 as PixelRatio,
  Platform,
  default40 as Pressable,
  default41 as RefreshControl,
  default42 as SafeAreaView,
  default43 as ScrollView,
  default32 as SectionList,
  default24 as Share,
  default44 as StatusBar,
  default27 as Switch,
  default45 as Text,
  default46 as TextInput,
  default30 as TouchableHighlight,
  default31 as TouchableNativeFeedback,
  default33 as TouchableOpacity,
  default34 as TouchableWithoutFeedback,
  UIManager,
  default25 as Vibration,
  default47 as View,
  default28 as VirtualizedList,
  canUseDOM,
  dismissKeyboard,
  findNodeHandle,
  isWebColor,
  mergeRefs,
  processColor,
  default3 as render,
  requireNativeComponent,
  unmountComponentAtNode,
  default2 as unstable_createElement,
  default49 as useColorScheme,
  useEvent,
  useHover,
  useLocaleContext,
  useMergeRefs,
  default50 as useWindowDimensions
};
