// react-native-web-lite/dist/esm/AccessibilityInfo/index.js
import { canUseDOM } from "react-native-web-internals";
function isScreenReaderEnabled() {
  return new Promise((resolve, reject) => {
    resolve(!0);
  });
}
const prefersReducedMotionMedia = canUseDOM && typeof window.matchMedia == "function" ? window.matchMedia("(prefers-reduced-motion: reduce)") : null;
function isReduceMotionEnabled() {
  return new Promise((resolve, reject) => {
    resolve(prefersReducedMotionMedia ? prefersReducedMotionMedia.matches : !0);
  });
}
function addChangeListener(fn) {
  prefersReducedMotionMedia != null && (prefersReducedMotionMedia.addEventListener != null ? prefersReducedMotionMedia.addEventListener("change", fn) : prefersReducedMotionMedia.addListener(fn));
}
function removeChangeListener(fn) {
  prefersReducedMotionMedia != null && (prefersReducedMotionMedia.removeEventListener != null ? prefersReducedMotionMedia.removeEventListener("change", fn) : prefersReducedMotionMedia.removeListener(fn));
}
const handlers = {}, AccessibilityInfo = {
  /**
   * Query whether a screen reader is currently enabled.
   *
   * Returns a promise which resolves to a boolean.
   * The result is `true` when a screen reader is enabled and `false` otherwise.
   */
  isScreenReaderEnabled,
  /**
   * Query whether the user prefers reduced motion.
   *
   * Returns a promise which resolves to a boolean.
   * The result is `true` when a screen reader is enabled and `false` otherwise.
   */
  isReduceMotionEnabled,
  /**
   * Deprecated
   */
  fetch: isScreenReaderEnabled,
  /**
   * Add an event handler. Supported events: reduceMotionChanged
   */
  addEventListener: function(eventName, handler) {
    if (eventName === "reduceMotionChanged") {
      if (!prefersReducedMotionMedia)
        return;
      const listener = (event) => {
        handler(event.matches);
      };
      addChangeListener(listener), handlers[handler] = listener;
    }
    return {
      remove: () => AccessibilityInfo.removeEventListener(eventName, handler)
    };
  },
  /**
   * Set accessibility focus to a react component.
   */
  setAccessibilityFocus: function(reactTag) {
  },
  /**
   * Post a string to be announced by the screen reader.
   */
  announceForAccessibility: function(announcement) {
  },
  /**
   * Remove an event handler.
   */
  removeEventListener: function(eventName, handler) {
    if (eventName === "reduceMotionChanged") {
      const listener = handlers[handler];
      if (!listener || !prefersReducedMotionMedia)
        return;
      removeChangeListener(listener);
    }
  }
};
var AccessibilityInfo_default = AccessibilityInfo;
export {
  AccessibilityInfo_default as default
};
