// react-native-web-lite/dist/cjs/TouchableWithoutFeedback.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf, __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: !0 }) : target,
  mod
)), __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var TouchableWithoutFeedback_exports = {};
__export(TouchableWithoutFeedback_exports, {
  default: () => TouchableWithoutFeedback_default
});
module.exports = __toCommonJS(TouchableWithoutFeedback_exports);
var React = __toESM(require("react")), import_react = require("react"), import_react_native_web_internals = require("react-native-web-internals");
const forwardPropsList = {
  accessibilityDisabled: !0,
  accessibilityLabel: !0,
  accessibilityLiveRegion: !0,
  accessibilityRole: !0,
  accessibilityState: !0,
  accessibilityValue: !0,
  children: !0,
  disabled: !0,
  focusable: !0,
  nativeID: !0,
  onBlur: !0,
  onFocus: !0,
  onLayout: !0,
  testID: !0
}, pickProps = (props) => (0, import_react_native_web_internals.pick)(props, forwardPropsList);
function TouchableWithoutFeedback(props, forwardedRef) {
  const {
    delayPressIn,
    delayPressOut,
    delayLongPress,
    disabled,
    focusable,
    onLongPress,
    onPress,
    onPressIn,
    onPressOut,
    rejectResponderTermination
  } = props, hostRef = (0, import_react.useRef)(null), pressConfig = (0, import_react.useMemo)(
    () => ({
      cancelable: !rejectResponderTermination,
      disabled,
      delayLongPress,
      delayPressStart: delayPressIn,
      delayPressEnd: delayPressOut,
      onLongPress,
      onPress,
      onPressStart: onPressIn,
      onPressEnd: onPressOut
    }),
    [
      disabled,
      delayPressIn,
      delayPressOut,
      delayLongPress,
      onLongPress,
      onPress,
      onPressIn,
      onPressOut,
      rejectResponderTermination
    ]
  ), pressEventHandlers = (0, import_react_native_web_internals.usePressEvents)(hostRef, pressConfig), element = React.Children.only(props.children), children = [element.props.children], supportedProps = pickProps(props);
  supportedProps.accessibilityDisabled = disabled, supportedProps.focusable = !disabled && focusable !== !1, supportedProps.ref = (0, import_react_native_web_internals.useMergeRefs)(forwardedRef, hostRef, element.ref);
  const elementProps = Object.assign(supportedProps, pressEventHandlers);
  return React.cloneElement(element, elementProps, ...children);
}
const MemoedTouchableWithoutFeedback = React.memo(
  React.forwardRef(TouchableWithoutFeedback)
);
MemoedTouchableWithoutFeedback.displayName = "TouchableWithoutFeedback";
var TouchableWithoutFeedback_default = MemoedTouchableWithoutFeedback;
